### PR TITLE
Footnotes

### LINKED ISSUE #56: Footnotes
Supported by, e.g. [pulldown-cmark](https://github.com/google/pulldown-cmark).

### PR DESCRIPTION
**Pull request by kivikakk**
Fixes #56.  Adds support for footnotes to `cmark-gfm`. This is a key requirement of being able to be used where Kramdown is; for instance, Kramdown is the default Markdown renderer in GitHub Pages.

This PR does the following:

* Generalises the map used for reference link lookups to support both reference links and footnotes.
  * It's still fairly specific to the needs of both (it's not a "generalised map" by any means), but that's fine.
* Parses footnote definitions (block-level) and references (inline-level) according to [kramdown's syntax](https://kramdown.gettalong.org/syntax.html#footnotes).
* At the parsing finalisation stage:
  * Collects definitions in one pass over the document and added to a map.
  * Collects references in another pass, assigning increasing indexes to the definitions so-referenced. (i.e. footnote #​1 is the first footnote referenced, not the first defined.)
  * Adds definitions that were referenced to the end of the AST in reference order.
* The HTML formatter produces links and footnote sections according to Kramdown's output format (for now).
* The CommonMark formatter roundtrips footnotes correctly.

---

To-do:

- [x] reduce duplication between reference map/footnote map
- [x] don't use linear search for footnote definitions when finalizing
- [ ] ~~fix the `static unsigned int ix;` hack in `commonmark.c`~~ whatever it's fine
- [x] take our `snprintf` results and turn them into asserts
- [x] make footnote parsing optional (it costs a bit)

### CODE REVIEW

**Comment by kivikakk** [comment]
Would love some review from @vmg and/or @charliesome.

File: src/commonmark.c [hunk 1]
```diff
@@ -162,6 +162,8 @@ static cmark_node *get_containing_block(cmark_node *node) {
   return NULL;
 }
 
+static unsigned int footnote_ix;
```
Comment (vmg) [inline src/commonmark.c]: Proooobably not fine. 😄

It seems like this single variable stops the whole library from being threadsafe. Can we give this some more thought? Surely the footnote index can fit in one of the many state structs we're already passing around.

Comment (kivikakk) [reply src/commonmark.c]: TOTALLY FINE.

(Shit, yeah.)

**Comment by metasean** [comment]
@kivikakk - What is the timeline for this footnote extension to be enabled in GitHub or for gists?

**Comment by kivikakk** [comment]
There are no plans to enable the extension on GitHub.com or in gists. This is for Pages only.

**Comment by metasean** [comment]
@kivikakk - Needless to say, this makes me really sad, but I really appreciate your definitive answer.

**Comment by vassudanagunta** [comment]
@metasean Good news: [Footnotes now supported in Markdown fields](https://github.blog/changelog/2021-09-30-footnotes-now-supported-in-markdown-fields/).

**Pull request merged**